### PR TITLE
Lab2 - BubbleSort and InsertionSort plus testing

### LINKED ISSUE #278: Lab 2 Bubble and Insertion Sort
Create an executable go program in directory 02_bubble/USERNAME
Write a function that returns a sorted copy of int slice s using Bubble sort
Bonus points: implement Insertion sort

### PR DESCRIPTION
**Pull request by Sarahdavy1983**
Fixes #278

Review of colleague's PR #261 

#### Changes proposed in this PR:
Create an executable go program in directory 02_bubble/USERNAME
Write a function that returns a sorted copy of int slice s using Bubble sort
Bonus points: implement Insertion sort

### CODE REVIEW

File: 02_bubble/sarahdavy1983/main_test.go
Comment (jessodri) [inline 02_bubble/sarahdavy1983/main_test.go:33]: Looks great to me!  My only suggestion would be to add a test case for if the array contains negative numbers :)

Comment (Sarahdavy1983) [reply 02_bubble/sarahdavy1983/main_test.go:33]: Thanks!!  This is now updated :)

File: 02_bubble/sarahdavy1983/main.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main.go]: Should use `var SliceCopy = s` or ` Slicecopy := s` to copy slice

File: 02_bubble/sarahdavy1983/main.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main.go]: same here

File: 02_bubble/sarahdavy1983/main.go
Comment (joshcarp) [reply 02_bubble/sarahdavy1983/main.go]: also there isn't any reason why you needed sliceCopy to start with; you can do it in place

File: 02_bubble/sarahdavy1983/main.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main.go]: I like the use of this; although to simplify we could use 
`for swapped := true; swapped;{ `if we wanted to condense a little more

File: 02_bubble/sarahdavy1983/main.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main.go]: this for loop can be consensed to declare j inline:
`for j := i; j > 0; {` 
Note that you might have tried this, but maybe just needed to add the trailing `;`

File: 02_bubble/sarahdavy1983/main_test.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main_test.go]: You're not testing anything here: you're going to need to capture the output

File: 02_bubble/sarahdavy1983/main_test.go
Comment (joshcarp) [inline 02_bubble/sarahdavy1983/main_test.go]: nitpick: tc could be changed with `test` as you're iterating through a list of `tests` and the singular of `tests` could be `test`
just gives more context to what `tc` is, although this is probablty really pedantic

**Review by joshcarp — CHANGES_REQUESTED** [review]
Nice!

File: 02_bubble/sarahdavy1983/main.go
Comment (Sarahdavy1983) [reply 02_bubble/sarahdavy1983/main.go]: @Joshcarp 
Thanks for the feedback, I have updated as per most of your comments.  I just want to clarify what you mean by " also there isn't any reason why you needed sliceCopy to start with; you can do it in place"?  Per my understanding of the requirements we need to create a copy

File: 02_bubble/sarahdavy1983/main.go
Comment (camscale) [inline 02_bubble/sarahdavy1983/main.go:34]: just a small nitpick: this should go in the `for` loop:
```
for j := i; j > 0; j-- {
...
}
```

File: 02_bubble/sarahdavy1983/main_test.go
Comment (camscale) [inline 02_bubble/sarahdavy1983/main_test.go:33]: I would suggest adding a test case for a 1-element slice and a 2-element slice. This is because "1" is a boundary condition for the length of the slice - you start counting at 1 and use `<`. If you get the logic slightly wrong in the code, it's easy to be off by one.

These off by one errors (boundary conditions) are common enough that you should look at the comparisons in your code and add test cases for either side of that boundary.

This is a general software engineering comment, not Go-specific, so this won't block approval if you would prefer to merge and move on. Just consider it for later labs.

**Review by camscale — CHANGES_REQUESTED** [review]
First, I am so sorry we've dropped the ball on reviewing your PR. The amount of time that this has been waiting for approval is way too long. We're working through the backlog and hope to keep the PR review wait time much shorter.

This code is very good. There's really only one little thing I would change which makes the insertion sort loop a little clearer. I'm keeping an eye on this so I'm ready to approve as soon as you push a new change. If you are on our slack channel (#go-course) you can ping me there too (camh) for a quick re-ewview.

**Comment by juliaogris** [comment]
The go-course is now closed. Thank you very much for participating.

**Pull request closed without merging**
